### PR TITLE
Introduce upgrade go-ahead and upgrade restriction signals

### PR DESCRIPTION
**Pull request by pepyakin**
This patch introduces a new mechanism that will be used in future by parachains to apply validation code upgrade.

Currently, the upgrades are assumed to be activated after a fixed number of relay-chain blocks. Also there are configuration values that the parachain should keep an eye on to not signal an upgrade before an upgrade cooldown (aka `validation_upgrade_frequency`) timer elapses. These mechanics do not work well with our planned changes such as https://github.com/paritytech/polkadot/issues/3211 and https://github.com/paritytech/polkadot/issues/3310.

Thus as a first step to tackle this problem this patch introduces two new signals that should coordinate an upgrade:

- upgrade go ahead signal
- upgrade restriction signal

Instead of counting the number of the relay chain blocks passed since the upgrade was signalled the parachain will be looking at the upgrade go ahead signal through merkle proofs. 

The logic of upgrades themselves is not changed at this time.

This signal is almost always absent. However, at the relay-chain block the upgrade expected to happen, this signal will be turned on. When a parachain block is created on top of that relay-parent or it's descendant, it will observe the go ahead signal and enact the upgrade. The next parablock will be validated using the upgraded validation code.

As soon as the first block with the activated upgrade is included, the signal is reset.

The upgrade restriction signal is set when validation code upgrades are disallowed. We do not give details on why at this moment. To follow the current upgrade logic, this signal is set as soon as the upgrade is announced. The signal is unset only after `validation_upgrade_frequency` is elapsed.

# Rationale

## Why two signals, it seems that it's possible to get away with only one?

It's indeed was initial thinking that one signal will do. 

When I approached this problem I realized that semantically those are different states and cannot be represented by a single state machine. A contrived example: We have a parathread. The parathread doesn't produce blocks too often. The parathread schedules an upgrade. Then the parathread produces a block after `validation_upgrade_frequency`. If the signal had a single state only we would have to signal GoAhead first. We have to delay signalling that the upgrade is allowed  to another relay-block.

Besides being semantically unsound, it was also harder to implement a single signal model.

### CODE REVIEW

File: roadmap/implementers-guide/src/runtime/paras.md
Comment (rphmeier) [inline roadmap/implementers-guide/src/runtime/paras.md:142]: typo: abscent -> absent

File: primitives/src/v1/mod.rs
Comment (rphmeier) [inline primitives/src/v1/mod.rs:1050]: It would be good to have explicit indices on these enums.

File: runtime/parachains/src/paras.rs
Comment (rphmeier) [inline runtime/parachains/src/paras.rs:670]: Should `UpgradeCooldowns` be pruned as well?

Comment (rphmeier) [reply runtime/parachains/src/paras.rs:670]: I see that it will be naturally pruned based on block, so it may be redundant.

**Review by rphmeier — APPROVED** [review]
Looks like a welcome refactoring and more elegant solution to the upgrade signal process. Is the plan to migrate Cumulus and eventually phase out some of the other storage members?

**Comment by pepyakin** [comment]
Yes, this patch is just a stepping stone. My rough plan is:

- merge this
- deploy this on {rococo, kusama}
- merge https://github.com/paritytech/cumulus/pull/517 and ask parachains to update their collators
- at this point, I imagine, `UpcomingUpgrades` won't be needed and will be substituted by another mechanism to signal the `GoAhead`. E.g. the pre-checking game.

File: runtime/parachains/src/paras.rs
Comment (pepyakin) [reply runtime/parachains/src/paras.rs:670]: Yeah, similar to `UpcomingUpgrades`. However, we perform insertions based on `UpcomingUpgrades` but for `UpgradeCooldowns` we only perform removals.

That said, I am ok to add filtering for cooldowns.

File: primitives/src/v1/mod.rs
Comment (pepyakin) [reply primitives/src/v1/mod.rs:1050]: https://github.com/paritytech/polkadot/pull/3371/commits/b058f8e6ff600d0a113bf21256f40c7b53b84267

File: roadmap/implementers-guide/src/runtime/paras.md
Comment (pepyakin) [reply roadmap/implementers-guide/src/runtime/paras.md]: https://github.com/paritytech/polkadot/pull/3371/commits/d77aec15e9f35454970d595aa2ce34059c4bf4a7

File: runtime/parachains/src/paras.rs
Comment (pepyakin) [reply runtime/parachains/src/paras.rs:673]: https://github.com/paritytech/polkadot/pull/3371/commits/d0c134e794682405a6374e0e293d66671e38803a

**Comment by rphmeier** [comment]
LGTM after merge addressed

**Pull request merged**
